extensions/sql/policy/store/src/main/java/org/eclipse/dataspaceconnector/sql/policy/store/SqlPolicyStore.java

```diff
@@ -0,0 +1,176 @@
+import org.eclipse.dataspaceconnector.spi.transaction.datasource.DataSourceRegistry;
+import org.eclipse.dataspaceconnector.spi.types.TypeManager;
+import org.jetbrains.annotations.Nullable;
+
+import javax.sql.DataSource;
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Stream;
+
+import static org.eclipse.dataspaceconnector.sql.SqlQueryExecutor.executeQuery;
+
+public class SqlPolicyStore implements PolicyStore {
+
+    private final DataSourceRegistry dataSourceRegistry;
+    private final String dataSourceName;
+    private final TransactionContext transactionContext;
+    private final TypeManager typeManager;
+    private final SqlPolicyStoreStatements sqlPolicyStoreStatements;
+
+    public SqlPolicyStore(DataSourceRegistry dataSourceRegistry, String dataSourceName, TransactionContext transactionContext, TypeManager typeManager, SqlPolicyStoreStatements sqlPolicyStoreStatements) {
+        this.dataSourceRegistry = Objects.requireNonNull(dataSourceRegistry);
+        this.dataSourceName = Objects.requireNonNull(dataSourceName);
+        this.transactionContext = Objects.requireNonNull(transactionContext);
+        this.typeManager = Objects.requireNonNull(typeManager);
+        this.sqlPolicyStoreStatements = Objects.requireNonNull(sqlPolicyStoreStatements);
+    }
+
+    @Override
+    public @Nullable Policy findById(String id) {
+        try {
+            return single(executeQuery(getConnection(), this::mapResultSet, sqlPolicyStoreStatements.getSqlFindByClauseTemplate(), id));
+        } catch (SQLException exception) {
+            throw new EdcPersistenceException(exception);
+        }
+    }
+
+    @Override
+    public Stream<Policy> findAll(QuerySpec spec) {
+        return transactionContext.execute(() -> {
+            Objects.requireNonNull(spec);
+
+            var limit = Limit.Builder.newInstance()
+                    .limit(spec.getLimit())
+                    .offset(spec.getOffset())
+                    .build();
+
+            var query = sqlPolicyStoreStatements.getSqlFindClauseTemplate() + " " + limit.getStatement();
+
+            try (var connection = getConnection()) {
+                var policies = executeQuery(
+                        connection,
+                        this::mapResultSet,
+                        String.format(query, sqlPolicyStoreStatements.getPolicyTableName()));
+                return policies.stream();
+            } catch (Exception exception) {
+                throw new EdcPersistenceException(exception);
+            }
+        });
+    }
+
+    @Override
+    public void save(Policy policy) {
+        Objects.requireNonNull(policy);
+
+        transactionContext.execute(() -> {
+            try (var connection = getConnection()) {
+                executeQuery(connection, sqlPolicyStoreStatements.getSqlSaveClauseTemplate(),
```

Comment:
I'd rather have an explicit "if-exists" check here (either using `SELECT` or `SELECT COUNT`), even if that's another SQL statement

Reply:
yeah, but relying on an exception for condition checking is bad style

Reply:
I have implemented two additional private methods update and insert. When a policy with same ID exists I do an update and when not I insert it.